internal/backends/local/local.go

```diff
@@ -0,0 +1,157 @@
+package local
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"strings"
+	"sync"
+
+	"github.com/cloudquery/plugin-sdk/specs"
+)
+
+type Local struct {
+	sourceName string
+	spec       Spec
+	tables     map[string]entries // table -> key -> value
+	tablesLock sync.RWMutex
+}
+
+type entries map[string]string
+
+func New(sourceSpec specs.Source) (*Local, error) {
+	spec := Spec{}
+	err := sourceSpec.UnmarshalBackendSpec(&spec)
+	if err != nil {
+		return nil, err
+	}
+	spec.SetDefaults()
+
+	l := &Local{
+		sourceName: sourceSpec.Name,
+		spec:       spec,
+	}
+	tables, err := l.loadPreviousState()
+	if err != nil {
+		return nil, err
+	}
+	if tables == nil {
+		tables = map[string]entries{}
+	}
+	l.tables = tables
+	return l, nil
+}
+
+func (l *Local) loadPreviousState() (map[string]entries, error) {
+	files, err := os.ReadDir(l.spec.Path)
+	if os.IsNotExist(err) {
+		return nil, nil
+	}
+	var tables = map[string]entries{}
+	for _, f := range files {
+		if f.IsDir() || !f.Type().IsRegular() {
+			continue
+		}
+		name := f.Name()
+		if !strings.HasSuffix(name, ".json") || !strings.HasPrefix(name, l.sourceName+"-") {
+			continue
+		}
+		table, kv, err := l.readFile(name)
+		if err != nil {
+			return nil, err
+		}
+		tables[table] = kv
+	}
+	return tables, nil
+}
+
+func (l *Local) readFile(name string) (table string, kv entries, err error) {
+	p := path.Join(l.spec.Path, name)
+	f, err := os.Open(p)
+	if err != nil {
+		return "", nil, fmt.Errorf("failed to open state file: %w", err)
+	}
+	b, err := io.ReadAll(f)
+	if err != nil {
+		return "", nil, fmt.Errorf("failed to read state file: %w", err)
+	}
+	err = f.Close()
+	if err != nil {
+		return "", nil, fmt.Errorf("failed to close state file: %w", err)
+	}
+	err = json.Unmarshal(b, &kv)
+	if err != nil {
+		return "", nil, fmt.Errorf("failed to unmarshal state file: %w", err)
+	}
+	table = strings.TrimPrefix(strings.TrimSuffix(name, ".json"), l.sourceName+"-")
+	return table, kv, nil
+}
+
+func (l *Local) Get(_ context.Context, table, key string) (string, error) {
+	l.tablesLock.RLock()
+	defer l.tablesLock.RUnlock()
+
+	if _, ok := l.tables[table]; !ok {
+		return "", nil
+	}
+	return l.tables[table][key], nil
+}
+
+func (l *Local) Set(_ context.Context, table, key, value string) error {
+	l.tablesLock.Lock()
+	defer l.tablesLock.Unlock()
```

Comment:
I think if you pre-alloc the tables you can avoid locking the whole map and just locking on a per table.

Reply:
Maybe I'm misunderstanding, but I think we can't pre-alloc tables, because we don't know which tables will use this before `Set` is called. And if no tables use it, we don't want to write any file when the backend gets closed.